PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc

```diff
@@ -188,8 +201,92 @@ PATMuonProducer::~PATMuonProducer()
 {
 }
 
+GlobalPoint* PATMuonProducer::getMuonDirection(const reco::MuonChamberMatch& chamberMatch,
+						const edm::ESHandle<GlobalTrackingGeometry>& geometry,
+						const DetId& chamberId)
+{
+  const GeomDet* chamberGeometry = geometry->idToDet( chamberId );
+  if (chamberGeometry){
+    LocalPoint localPosition(chamberMatch.x, chamberMatch.y, 0);
+    return new GlobalPoint(chamberGeometry->toGlobal(localPosition));
+  }
+  return nullptr;
+
+}
+
+
+void PATMuonProducer::fillL1TriggerInfo(pat::Muon& aMuon,
+					edm::Handle<std::vector<pat::TriggerObjectStandAlone> >& triggerObjects,
+					const edm::TriggerNames & names,
+					const edm::ESHandle<GlobalTrackingGeometry>& geometry)
+{
+  // L1 trigger object parameters are defined at MB2/ME2. Use the muon
+  // chamber matching information to get the local direction of the
+  // muon trajectory and convert it to a global direction to match the
+  // trigger objects
+
+  GlobalPoint* muonPosition(nullptr);
+  // loop over chambers
+  for ( const auto& chamberMatch: aMuon.matches() ) {
+    if ( chamberMatch.id.subdetId() == MuonSubdetId::DT) {
+      DTChamberId detId(chamberMatch.id.rawId());
+      if (abs(detId.station())!=2) continue;
+      muonPosition = getMuonDirection(chamberMatch, geometry, detId);
+      break;
+    }
+    if ( chamberMatch.id.subdetId() == MuonSubdetId::CSC) {
+      CSCDetId detId(chamberMatch.id.rawId());
+      if (abs(detId.station())!=2) continue;
+      muonPosition = getMuonDirection(chamberMatch, geometry, detId);
+      break;
+    }
+  }
+  if (not muonPosition) return;
+  for (auto triggerObject: *triggerObjects){
+    if (triggerObject.hasTriggerObjectType(trigger::TriggerL1Mu)){
+      if (deltaR(triggerObject.p4(),*muonPosition)>0.1) continue;
+      pat::TriggerObjectStandAlone obj(triggerObject);
+      obj.unpackPathNames(names);
+      aMuon.addTriggerObjectMatch(obj);
+    }
+  }
+  if (muonPosition) delete muonPosition;
+}
+
+void PATMuonProducer::fillHltTriggerInfo(pat::Muon& muon,
+					 edm::Handle<std::vector<pat::TriggerObjectStandAlone> >& triggerObjects,
+					 const edm::TriggerNames & names,
+					 const std::vector<std::string>& collection_filter_names)
+{
+  // WARNING: in a case of close-by muons the dR matching may select both muons.
+  // It's better to select the best match for a given collection.
+  for (auto triggerObject: *triggerObjects){
```

Comment:
this means the triggerObject is a copy. It looks like ```const auto&``` will work to avoid copies

Reply:
fixed